packages/dds/tree/src/simple-tree/tree.ts

```diff
@@ -140,18 +161,117 @@ export class TreeViewConfiguration<TSchema extends ImplicitFieldSchema = Implici
+				getOrAddEmptyToMap(allObjectKeys, key).push(schema);
+			}
+		} else if (schema.kind === NodeKind.Array) {
+			arrays.push(schema);
+		} else {
+			assert(schema.kind === NodeKind.Map, "invalid schema");
+			maps.push(schema);
+		}
+	}
+
+	if (arrays.length > 1) {
+		errors.push(
+			`More than one kind of array allowed within union (${JSON.stringify(arrays.map((s) => s.identifier))}). This would require type disambiguation which is not supported by arrays during import or export to JSON.`,
+		);
+	}
+
+	if (maps.length > 1) {
+		errors.push(
+			`More than one kind of map allowed within union (${JSON.stringify(maps.map((s) => s.identifier))}). This would require type disambiguation which is not supported by maps during import or export to JSON.`,
+		);
+	}
+
+	if (maps.length > 0 && arrays.length > 0) {
+		errors.push(
+			`Both a map and an array allowed within union (${JSON.stringify([...arrays, ...maps].map((s) => s.identifier))}). Both can be implicitly constructed from arrays, which are ambiguous when the array is empty.`,
+		);
+	}
+
+	if (objects.length > 0 && maps.length > 0) {
+		errors.push(
+			`Both a object and a map allowed within union (${JSON.stringify([...objects, ...maps].map((s) => s.identifier))}). Support for constructing maps from objects is planned this will become ambiguous when that is supported.`,
+		);
+	}
+
+	// Check for objects which fully overlap:
+	for (const schema of objects) {
+		// Check that each object has at least one required field which is unique to it.
+		// This is a conservative uniqueness check that ensures that its possible to disambiguate objects,
+		// even if they have extra non-field members on them (that don't collide with any fields in the union), without even checking the types of fields.
+		let hasUniqueField = false;
+		for (const [key, field] of schema.fields) {
+			if (field.kind === FieldKind.Required) {
+				const withKey = allObjectKeys.get(key) ?? fail("missing schema");
+				if (withKey.length === 0) {
+					hasUniqueField = true;
+				}
+			}
+		}
+
+		if (hasUniqueField === false) {
+			// TODO: make this check more permissive.
+			// Allow using the type of the field to disambiguate, at least for leaf types.
+			// Add "constant" fields which can be used to disambiguate even more cases without adding persisted data: maybe make them optional in constructor?
+			// Consider separating unambiguous implicit construction format from constructor arguments at type level, allowing constructor to superset the implicit construction options (ex: optional constant fields).
+			// The policy here however must remain at least as conservative as shallowCompatibilityTest in src/simple-tree/toMapTree.ts.
+
+			errors.push(
+				`Object ${JSON.stringify(schema.identifier)} does not have a unique required field within union (${JSON.stringify([...union].map((s) => s.identifier))}). Support for constructing maps from objects is planned this will become ambiguous when that is supported.`,
```

Comment:
Nit: I'm not sure the note about future map construction support is needed. Just noting that there are no unique required fields, and that leads to (potential) ambiguity is probably sufficient.

Reply:
If maps are required to be constructed from iterables (like currently), then non iterable objects are unambiguously objects and not maps.

Reply:
My point is more that end-users don't need to know about our future plans. We can just inform them of what is not supported and what to do instead.

Reply:
Probably better (if desired) to just leave a comment in the code about our future plans.